fire/cli/niv/__init__.py

```diff
@@ -500,16 +500,79 @@ def find_sagsid(sagsgang: pd.DataFrame) -> str:
 def punkter_geojson(
     projektnavn: str,
     punkter: pd.DataFrame,
+    præfix: str = "",
 ) -> None:
     """Skriv punkter/koordinater i geojson-format"""
-    with open(f"{projektnavn}-punkter.geojson", "wt") as punktfil:
+    with open(f"{projektnavn}{præfix}-punkter.geojson", "wt") as punktfil:
         til_json = {
             "type": "FeatureCollection",
             "Features": list(punkt_feature(punkter)),
         }
         json.dump(til_json, punktfil, indent=4)
 
 
+# ------------------------------------------------------------------------------
+def obs_feature(
+    punkter: pd.DataFrame, observationer: pd.DataFrame, antal_målinger: Dict[Tuple, int]
+) -> Dict[str, str]:
+    """Omsæt observationsinformationer til JSON-egnet dict"""
+    for i in range(observationer.shape[0]):
+        fra = observationer.at[i, "Fra"]
+        til = observationer.at[i, "Til"]
+        feature = {
+            "type": "Feature",
+            "properties": {
+                "Fra": fra,
+                "Til": til,
+                "Målinger": antal_målinger[tuple(sorted([fra, til]))],
+                "Afstand": observationer.at[i, "L"],
+                "ΔH": observationer.at[i, "ΔH"],
+                # konvertering, da json.dump ikke uderstøtter int64
+                "Opstillinger": int(observationer.at[i, "Opst"]),
+                "Journal": observationer.at[i, "Journal"],
+                "Type": observationer.at[i, "Type"],
+                "Slukket": observationer.at[i, "Sluk"],
+            },
+            "geometry": {
+                "type": "LineString",
+                "coordinates": [
+                    [punkter.at[fra, "Øst"], punkter.at[fra, "Nord"]],
+                    [punkter.at[til, "Øst"], punkter.at[til, "Nord"]],
+                ],
+            },
+        }
+        yield feature
+
+
+# ------------------------------------------------------------------------------
+def observationer_geojson(
+    projektnavn: str,
+    punkter: pd.DataFrame,
+    observationer: pd.DataFrame,
+    præfix: str = "",
+) -> None:
+    """Skriv observationer til geojson-fil"""
+
+    fra = observationer["Fra"]
+    til = observationer["Til"]
+
+    # Optæl antal frem-og/eller-tilbagemålinger pr. strækning: Vi starter
+    # med en dict med et nul for hver strækning
+    par = [tuple(p) for p in zip(fra, til)]
+    antal_målinger = dict((tuple(sorted(p)), 0) for p in par)
+    # ...og så tæller vi det relevante element op for hver observation
+    for p in par:
+        # Indeksering med tuple(sorted(p)) da set(p) ikke kan hashes
+        antal_målinger[tuple(sorted(p))] += 1
+
+    with open(f"{projektnavn}{præfix}-observationer.geojson", "wt") as obsfil:
+        til_json = {
+            "type": "FeatureCollection",
+            "Features": list(obs_feature(punkter, observationer, antal_målinger)),
+        }
+        json.dump(til_json, obsfil, indent=4)
```

Comment:
Her binder vi os til at skrive den resulterende json-struktur direkte til en fil med et bestemt filnavn. Det gør det svært at genbruge funktionaliteten, der producerer json-strukturen et sted, hvor det ikke skal gemmes osm fil, men måske skal sammenknyttes med andet eller sendes via en service. Altså andet I/O. Jeg foreslår, at IO-delen må være op til dén del af koden, der kalder funktionen.

Reply:
Hvordan tænker du præcist her? Funktionen returnerer en geojson-streng skabt med `json.dumps()` eller dicten `til_json`?

Jeg kan godt se fidusen i at skille tingene ad men er tilbøjelig til at sige at er noget vi kan gøre hvis behøver viser sig på et senere tidspunkt.

Reply:
I forhold til brugen af den kan det godt vente til, vi har behov for andet. Men i forhold til test af funktionaliten med at opbygge JSON-strukturen, så er det nemmere at teste, at den gør, hvad vi forventer, hvis IO-delen bliver separeret fra funktionaliteten med at bygge indholdet, der skal gemmes (eller eksempelvis transmitteres videre). Det er også nemmere at skrive tests til simpel ting, end at gøre det senere, hvis vi har udvidet dem.

Men så er der funktionens navn, som indeholder ordet geojson. Hvis vi vælger at skille IO og JSON-struktur, så skal funktionen returnere noget, vi betragter som gyldig geojson. Skal det være en Python `dict` eller den endelige GeoJSON-tekst, som kan fås med `json.dumps`? Det kommer an på, hvordan det nuværende navn tolkes.

Det er bedst, hvis funktionen returnerer dét, dens navn antyder. Så hvis den laver GeoJSON, så burde det vel egentlig være tekst, den returnerer. Så vil en test skulle sikre, at det er korrekt GeoJSON, og/eller at data bliver korrekt oversat til formatet. Hvis vi kun skal teste struktur og ikke korrekt dataoversættelse til tekst-formatet, så er det nemmere at have funktionen returnere en `dict`, og så checke, at den bliver oprettet som forventet. En GeoJSON-tekst-validering ville være god at have, men indtil videre er det vel også nok, at vi kan validere med det samme, om formatet er korrekt, i QGIS eller lignende:)